### PR TITLE
Handle synchronisation of generateTree method at the "lowest" level

### PR DESCRIPTION
**Pull request by dordsor21**
- https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/1788

### CODE REVIEW

**Review by NotMyFault — APPROVED** [review]

File: spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java [hunk 1]
```diff
@@ -572,27 +573,32 @@ public boolean generateTree(
...
+        final BlockVector3 finalBlockVector = pt;
+        // Sync to main thread to ensure no clashes occur
+        Map<BlockPosition, CraftBlockState> placed = TaskManager.taskManager().sync(() -> {
+            world.captureTreeGeneration = true;
+            world.captureBlockStates = true;
+            boolean grownTree = bukkitWorld.generateTree(BukkitAdapter.adapt(bukkitWorld, finalBlockVector), bukkitType);
+            world.captureBlockStates = false;
+            world.captureTreeGeneration = false;
+            if (!grownTree) {
+                world.capturedBlockStates.clear();
```
Comment (SirYwell) [inline spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java:585]: I think the map should always be cleared (but copied before when returning)

File: spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java [hunk 2]
```diff
@@ -572,27 +573,32 @@ public boolean generateTree(
...
-        } else {
-            for (CraftBlockState craftBlockState : world.capturedBlockStates.values()) {
-                if (craftBlockState == null || craftBlockState.getType() == Material.AIR) {
-                    continue;
-                }
-                editSession.setBlock(craftBlockState.getX(), craftBlockState.getY(), craftBlockState.getZ(),
-                        BukkitAdapter.adapt(((org.bukkit.block.BlockState) craftBlockState).getBlockData())
-                );
+        }
+        for (CraftBlockState craftBlockState : world.capturedBlockStates.values()) {
```
Comment (SirYwell) [inline spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java:593]: This should use the copied map (or we could also just copy the values instead)

**Review by SirYwell — CHANGES_REQUESTED** [review]
The comments apply for all versions I guess

File: spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java [hunk 3]
```diff
@@ -572,27 +573,32 @@ public boolean generateTree(
...
+            try {
+                bukkitWorld.generateTree(BukkitAdapter.adapt(bukkitWorld, finalBlockVector), bukkitType);
+                return ImmutableMap.copyOf(world.capturedBlockStates);
+            } finally {
+                world.captureBlockStates = false;
+                world.captureTreeGeneration = false;
+                world.capturedBlockStates.clear();
+            }
+        });
+        if (placed == null) {
```
Comment (SirYwell) [inline spigot_v1_16_R3/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/FAWE_Spigot_v1_16_R3.java:590]: Without checking the return value of generateTree, placed won't be null.

**Review by SirYwell — CHANGES_REQUESTED** [review]

**Review by SirYwell — APPROVED** [review]

**Pull request merged**